R/epiparameter.R

```diff
@@ -0,0 +1,996 @@
+      warning(
+        "Discretising a truncated continuous distribution, ",
+        "returning non-truncated discretised distribution",
+        call. = FALSE
+      )
+
+      idx <- grep(
+        pattern = "lower|upper",
+        x = names(prob_dist_params),
+        ignore.case = TRUE
+      )
+      prob_dist_params <- prob_dist_params[-idx]
+
+      # trunc dist family is truncated so get prob dist by unclassing dist and
+      # extracting name
+      list_dist <- unclass(x$prob_distribution)
+      prob_dist <- gsub(
+        pattern = "dist_",
+        replacement = "",
+        x = class(list_dist[[1]][[1]])[1],
+        fixed = TRUE
+      )
+    }
+
+    # standardise distribution parameter names
+    prob_dist_params <- .clean_params(
+      prob_distribution = prob_dist,
+      prob_distribution_params = prob_dist_params
+    )
+
+    # create a new discretised probability distribution
+    x$prob_distribution <- create_prob_distribution(
+      prob_distribution = prob_dist,
+      prob_distribution_params = prob_dist_params,
+      discretise = TRUE,
+      truncation = NA
+    )
+  }
+
+  # return epiparameter
+  x
+}
+
+#' @rdname discretise
+#' @export
+discretise.default <- function(x, ...) {
+  stop("No discretise method defined for class ", class(x))
+}
+
+#' Family method for the `<epiparameter>` class
+#'
+#' @description The [family()] function is used to extract the distribution
+#' names from objects from `{distributional}` and `{distcrete}`. This method
+#' provides the same interface for `<epiparameter>` objects to give consistent
+#' output irrespective of the internal distribution class.
+#'
+#' @param object An `<epiparameter>` object.
+#' @inheritParams stats::family
+#'
+#' @return A character string with the name of the distribution, or `NA` when
+#' the `<epiparameter>` object is unparameterised.
+#' @importFrom stats family
+#' @export
+#'
+#' @examples
+#' # example with continuous distribution
+#' ep <- epiparameter(
+#'   disease = "ebola",
+#'   epi_name = "incubation_period",
+#'   prob_distribution = create_prob_distribution(
+#'     prob_distribution = "gamma",
+#'     prob_distribution_params = c(shape = 1, scale = 1)
+#'   )
+#' )
+#' family(ep)
+#'
+#' # example with discretised distribution
+#' ep <- epiparameter(
+#'   disease = "ebola",
+#'   epi_name = "incubation_period",
+#'   prob_distribution = create_prob_distribution(
+#'     prob_distribution = "lnorm",
+#'     prob_distribution_params = c(meanlog = 1, sdlog = 1),
+#'     discretise = TRUE
+#'   )
+#' )
+#' family(ep)
+family.epiparameter <- function(object, ...) {
+  if (inherits(object$prob_distribution, "distcrete")) {
+    prob_dist <- object$prob_distribution$name
+  } else if (inherits(object$prob_distribution, "distribution")) {
+    if (is_truncated(object)) {
+      prob_dist <- gsub(
+        pattern = "dist_",
+        replacement = "",
+        x = class(unclass(unclass(object$prob_distribution)[[1]])[[1]])[1],
```

Comment:
Similar to the previous comment on the complicated indexing, might benefit from a comment.

Reply:
Thanks both for this suggestion. I've made improvements to these parts of the code in PR #398 by generalising `.get_mixture_dist()` to `.distributional_family()` which enables the access of the distribution names, for both transformed (e.g. mixture or truncated) or untransformed (e.g. gamma or lognormal) distributions. The `.distributional_family()` took inspiration from some code posted on mitchelloharawild/distributional#110 and as stated on that issue if a new function gets added to {distributional} we can potentially remove some code from this package and rely on their exported function.